### PR TITLE
Destroy command

### PR DESCRIPTION
**Pull request by phanimarupaka**
*Why*

This PR is for the destroy command. Will add more tests and status assertions once status PR is merged.

### CODE REVIEW

File: cmd/destroy/cmddestroy.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+			}
+
+			cmdutil.CheckErr(applier.Initialize(cmd))
+
+			// Create a context with the provided timout from the cobra parameter.
+			ctx, cancel := context.WithTimeout(context.Background(), applier.StatusOptions.Timeout)
+			defer cancel()
+			// Run the applier. It will return a channel where we can receive updates
+			// to keep track of progress and any issues.
+			ch := applier.Run(ctx)
```
Comment (mortent) [inline cmd/destroy/cmddestroy.go]: I'm not sure we should use the applier for this.

File: cmd/destroy/cmddestroy.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	"k8s.io/kubectl/pkg/cmd/util"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/util/i18n"
+	"sigs.k8s.io/cli-utils/pkg/apply"
+)
+
+// NewCmdDestroy creates the `destroy` command
+func NewCmdDestroy(f util.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	applier := apply.NewApplier(f, ioStreams)
+	applier.PruneOptions.Destroy = true
```
Comment (mortent) [inline cmd/destroy/cmddestroy.go]: This doesn't seem like a good way to configure it. In particular setting fields multiple levels down.

File: cmd/main.go [hunk 3]
```diff
@@ -6,6 +6,7 @@ package main
 import (
 	"flag"
 	"os"
+	"sigs.k8s.io/cli-utils/cmd/destroy"
```
Comment (mortent) [inline cmd/main.go]: This import should not be in this section.

File: cmd/main.go [hunk 4]
```diff
@@ -41,13 +42,15 @@ func main() {
...
 	}
 
-	names := []string{"apply", "diff"}
+	names := []string{"apply", "diff", "destroy"}
 	applyCmd := apply.NewCmdApply(f, ioStreams)
 	updateHelp(names, applyCmd)
 	diffCmd := diff.NewCmdDiff(f, ioStreams)
 	updateHelp(names, diffCmd)
+	destroyCmd := destroy.NewCmdDestroy(f, ioStreams)
+	updateHelp(names, destroyCmd)
```
Comment (mortent) [inline cmd/main.go]: I don't think this call is needed.

File: pkg/apply/applier.go [hunk 5]
```diff
@@ -140,7 +140,13 @@ func (a *Applier) Run(ctx context.Context) <-chan Event {
 		// This provides us with a slice of all the objects that will be
 		// applied to the cluster.
 		infos, _ := a.ApplyOptions.GetObjects()
-		err := a.ApplyOptions.Run()
+		var err error
+		// apply only if Destroy flag is off. Else skip the apply step and proceed to prune.
+		if !a.PruneOptions.Destroy {
+			err = a.ApplyOptions.Run()
+		} else {
+			err = nil
```
Comment (mortent) [inline pkg/apply/applier.go]: This is not needed. `err` is already set to nil by default on line 143

File: cmd/destroy/cmddestroy.go (hunk 2 above)
Comment (phanimarupaka) [reply cmd/destroy/cmddestroy.go]: Yeah. This is something I wanted to ask. Is applier a right place to add this field? Like applier.Destroy ?

File: cmd/main.go (hunk 3 above)
Comment (phanimarupaka) [reply cmd/main.go]: Will do

File: cmd/main.go (hunk 4 above)
Comment (phanimarupaka) [reply cmd/main.go]: Will do.

File: pkg/apply/applier.go (hunk 5 above)
Comment (phanimarupaka) [reply pkg/apply/applier.go]: I wanted it to be explicit and call it out so that it's more readable.

File: cmd/destroy/cmddestroy.go (hunk 2 above)
Comment (seans3) [reply cmd/destroy/cmddestroy.go]: We may not need a `Destroy` field anywhere.

File: pkg/apply/prune/prune.go [hunk 6]
```diff
@@ -48,6 +48,7 @@ type PruneOptions struct {
 
 	DryRun    bool
 	validator validation.Schema
+	Destroy   bool
```
Comment (seans3) [inline pkg/apply/prune/prune.go]: I don't think we need a `Destroy` boolean in the `prune`.

File: cmd/destroy/cmddestroy.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	printer := &apply.BasicPrinter{
+		IOStreams: ioStreams,
+	}
+
+	cmd := &cobra.Command{
+		Use:                   "destroy (-f FILENAME | -k DIRECTORY)",
+		DisableFlagsInUseLine: true,
+		Short:                 i18n.T("Destroy all the resources related to configuration managed by kpt"),
```
Comment (barney-s) [inline cmd/destroy/cmddestroy.go]: s/kpt/kapply ?

File: pkg/apply/destroyer.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	return ch
+}
+
+// SetFlags configures the command line flags needed for destroy
+// This is a temporary solution as we should separate the configuration
+// of cobra flags from the Destroyer.
+func (d *Destroyer) SetFlags(cmd *cobra.Command) {
+	d.ApplyOptions.DeleteFlags.AddFlags(cmd)
+	d.ApplyOptions.RecordFlags.AddFlags(cmd)
+	d.ApplyOptions.PrintFlags.AddFlags(cmd)
```
Comment (mortent) [inline pkg/apply/destroyer.go]: We don't want to set the print flags. Since we don't use the print functionality from Kubectl, we currently ignore them.

**Comment by phanimarupaka** [comment]
@seans3 I tried to write some unit tests but it seems like destroyer is only a wrapper on to of apply options and prune options and has methods, Initialize and run which by themselves are dependent on other options. I feel e2e tests makes more sense and not unit tests. Even in applier_test, run and initialize methods are not unit tested. Let me know your thoughts.

**Comment by mortent** [comment]
We shouldn't use e2e tests as a replacement for unit tests. If it is hard to write unit tests (and I can see that unit tests for the Destroyer might be hard right now), we need to clean things up to make it easier.

**Comment by mortent** [comment]
I'm ok with merging this without the unit tests. Those will require some work and perhaps upstream changes.
/lgtm

**Pull request merged**